queryserver-orchestrator/src/main/java/org/apache/phoenix/queryserver/orchestrator/TestExecutorClient.java

```diff
@@ -0,0 +1,122 @@
+import org.apache.curator.framework.recipes.leader.LeaderSelector;
+import org.apache.curator.framework.recipes.leader.LeaderSelectorListenerAdapter;
+import org.apache.curator.framework.state.ConnectionState;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+public class TestExecutorClient extends LeaderSelectorListenerAdapter implements Closeable {
+
+    private static final Logger logger = LoggerFactory.getLogger(TestExecutorClient.class);
+    private final LeaderSelector leaderSelector;
+
+    private Map<String, String> params;
+    private  ToolWrapper toolWrapper;
+
+    public TestExecutorClient(CuratorFramework client, String path, Map<String, String> params, ToolWrapper tool) {
+
+        // create a leader selector using the given path for management
+        // all participants in a given leader selection must use the same path
+        // ExampleClient here is also a LeaderSelectorListener but this isn't required
+        this.leaderSelector = new LeaderSelector(client, path, this);
+
+        // for most cases you will want your instance to requeue when it relinquishes leadership
+        this.leaderSelector.autoRequeue();
+
+        this.params = params;
+        this.toolWrapper = tool;
+    }
+
+    public void start() {
+        // the selection for this instance doesn't start until the leader selector is started
+        // leader selection is done in the background so this call to leaderSelector.start()
+        // returns immediately
+        leaderSelector.start();
+    }
+
+    public void close() {
+        leaderSelector.close();
+    }
+
+    @Override
+    public void takeLeadership(CuratorFramework client) throws Exception {
+        // we are now the leader. This method should not return until we want to relinquish
+        // leadership
+        logger.info("Took leadership.");
+
+        while (true) {
+            logger.info("Starting test case suite execution.");
+
+            executeQueryServerCanaryTool();
+
+            logger.info("Test suite execution completed. Waiting for " + params.get("interval")
+                    + " secs before executing next run.");
+
+            TimeUnit.SECONDS.sleep(Integer.parseInt(params.get("interval")));
+        }
+    }
+
+    @Override
+    public void stateChanged(CuratorFramework client, ConnectionState newState) {
+        logger.info("ZK Connection State Changed to [{}]", newState.name());
+        switch(newState) {
+            case CONNECTED:
+                logger.info("The host became the leader at " +System.currentTimeMillis());
+                break;
+            case LOST:
+                logger.info("The host lost leadership at "+ System.currentTimeMillis());
+                throw new CancelLeadershipException();
```

Comment:
This is likely going to be thrown in a separate thread. You need to interrupt the thread which is running your tool.

Reply:
yet to be addressed.

Reply:
Looks like this isn't addressed yet. The code may still continue running in split brain scenarios. I would suggest splitting the PR into two parts, and submit the leader election part later. How do you want to proceed?

Reply:
Good point, @karanmehta93 I ll add a TODO and point to new jira to handle this scenario. This needs extra attention.